pkg/cloud/api/copy.go

```diff
@@ -0,0 +1,303 @@
+		return fmt.Errorf("cannot set dest (%s)", p)
+	}
+	dest.Set(src)
+	cc.logS("copyBasic", "path", p, "value", dest.Interface())
+	return nil
+}
+
+func copyPointer(cc *copyContext, p Path, dest, src reflect.Value) error {
+	if dest.Type().Kind() != reflect.Pointer || src.Type().Kind() != reflect.Pointer {
+		return fmt.Errorf("copyPointer: invalid types: src %T, dest %T", src.Interface(), dest.Interface())
+	}
+	if src.IsZero() {
+		cc.logS("copyPointer zero", "path", p)
+		if !dest.CanSet() {
+			return fmt.Errorf("cannot set dest (%s)", p)
+		}
+		dest.Set(reflect.Zero(dest.Type()))
+		return nil
+	}
+	if dest.IsZero() {
+		if !dest.CanAddr() {
+			return fmt.Errorf("copyPointer: dest is nil and not addressable: src %T, dest %T", src.Interface(), dest.Interface())
+		}
+		dest.Set(reflect.New(dest.Type().Elem()))
+		cc.logS("copyPointer", "path", p, "value", dest.Interface())
+	}
+	return copyValuesImpl(cc, p.Pointer(), dest.Elem(), src.Elem())
+}
+
+func copySlice(cc *copyContext, p Path, dest, src reflect.Value) error {
+	if dest.Type().Kind() != reflect.Slice || src.Type().Kind() != reflect.Slice {
+		return fmt.Errorf("copySlice: invalid type (dest: %T, src: %T)", dest.Interface(), src.Interface())
+	}
+	if src.IsZero() {
+		dest.Set(reflect.Zero(dest.Type()))
+		cc.logS("copySlice zero", "path", p)
+		return nil
+	}
+
+	newSlice := reflect.MakeSlice(dest.Type(), src.Len(), src.Len())
+	for i := 0; i < src.Len(); i++ {
+		if err := copyValuesImpl(cc, p.Index(i), newSlice.Index(i), src.Index(i)); err != nil {
+			return err
+		}
+	}
+	cc.logS("copySlice", "path", p, "value", newSlice)
+
+	if !dest.CanSet() {
+		return fmt.Errorf("cannot set dest (%s)", p)
+	}
+	dest.Set(newSlice)
+
+	return nil
+}
+
+func copyStruct(cc *copyContext, p Path, dest, src reflect.Value) error {
+	if dest.Kind() != reflect.Struct || src.Kind() != reflect.Struct {
+		return fmt.Errorf("copyStruct: invalid type (dest: %T, src: %T)", dest.Interface(), src.Interface())
+	}
+	// Copy over fields that are present in both src and dest. Fields in dest
+	// that don't exist in src are left alone.
+	for i := 0; i < src.Type().NumField(); i++ {
+		srcFieldT := src.Type().Field(i)
+		fieldName := srcFieldT.Name
+		destField := dest.FieldByName(fieldName)
+		_, ok := dest.Type().FieldByName(fieldName)
+
+		if !ok {
+			// Only non-zero fields are counted towards
+			// the missing fields. Fields explicitly named
+			// in NullFields or ForceSendFields are
+			// handled by copyMetaFields() below.
+			if !src.Field(i).IsZero() {
+				cc.missing = append(cc.missing, missingFieldOnCopy{
+					Path:  p.Field(fieldName),
+					Value: src.Field(i).Interface(),
+				})
+				cc.logS("copyStruct missing field", "path", p, "fieldName", fieldName)
+			}
+			continue
+		}
+
+		// ServerResponse should be skipped.
+		if (p.Equal(Path{}) || p.Equal(Path{}.Pointer())) && fieldName == "ServerResponse" {
+			continue
+		}
+
+		if fieldName == "NullFields" || fieldName == "ForceSendFields" {
```

Comment:
do we need to associate these fields with a type?

Reply:
Are you saying we should also check the type in case there is a field with the name but it's not actually a metafield?

As far as I can tell -- these are always generated in every struct level.